File: lunabot_nav/scripts/dstar_node.py
#!/usr/bin/env python3

import numpy as np
import rospy
from geometry_msgs.msg import PoseStamped
from map_msgs.msg import OccupancyGridUpdate
from nav_msgs.msg import OccupancyGrid, Odometry, Path
from tf.transformations import quaternion_from_euler

from threading import Thread
from threading import Lock

from lunabot_nav.dstar import Dstar

class DstarNode:

    def __init__(self):
        self.map: np.ndarray[int] = np.array([]) # 2d array occupancy map: Occupancy probabilities from [0 to 100].  Unknown is -1.
        self.pose: list[float] = [] # [x, y], in meters/odom frame
        self.goal: list[float] = [] 

        self.grid_update_needed: bool = False
        self.goal_update_needed: bool = False

        self.resolution: float = 0  # meters per grid cell

        self.x_offset: float = 0 # real world pose of the point 0,0 in the grid
        self.y_offset: float = 0

        self.dstar: Dstar = None

        self.map_lock = Lock()

        self.path_sampling_rate = rospy.get_param("/nav/dstar_node/path_sampling_rate") # Take every <n-th> point from the path

        path_topic = rospy.get_param("/nav/global_path_topic")
        self.path_publisher = rospy.Publisher(path_topic, Path, queue_size=10, latch=True)

        self.occupancy_threshold = rospy.get_param("/nav/occ_threshold")



    def grid_callback(self, data: OccupancyGrid):
        """
        Update the map given a new occupancy grid. Update the flag such that dstar will update the map.
        """

        self.map_lock.acquire()

        rospy.loginfo("[Dstar node]: Got map")

        data_arr = np.array(data.data)

        width = data.info.width
        height = data.info.height

        if (np.all(data_arr == 0)): # ignore blank maps, still fix the height/width

            prev_x_offset = self.x_offset
            prev_y_offset = self.y_offset
    
            self.x_offset = data.info.origin.position.x
            self.y_offset = data.info.origin.position.y

            x_diff = data.info.origin.position.x - prev_x_offset
            y_diff = data.info.origin.position.y - prev_y_offset

            # if we get a blank map of a different side, keep the old map in place and pad it with -1 (unknown)
            # if the offset gets more negative, it means the map has expanded on the [left/top], otherwise on the [right/bottom]
            row_diff = data.info.height - self.map.shape[0]
            column_diff = data.info.width - self.map.shape[1]

            new_map = self.map

            if (row_diff != 0):
                new_rows = -1 * np.ones((row_diff, new_map.shape[1]))

                if (y_diff < 0):
                    # pad the top
                    new_map = np.concatenate((new_rows, new_map), axis=0)
                else:
                    # pad the bottom
                    new_map = np.concatenate((new_map, new_rows), axis=0)
            
            if (column_diff != 0):
                new_columns = -1 * np.ones((new_map.shape[0], column_diff))
                if (x_diff < 0):
                    # pad the left
                    new_map = np.concatenate((new_columns, new_map), axis=1)
                else:
                    # pad the right
                    new_map = np.concatenate((new_map, new_columns), axis=1)

            self.map = new_map
            self.map_lock.release()
            return

        self.map = np.reshape(data_arr, (height, width))

        self.resolution = data.info.resolution
        self.x_offset = data.info.origin.position.x
        self.y_offset = data.info.origin.position.y
        self.grid_update_needed = True

        self.map_lock.release()


    def grid_update_callback(self, data: OccupancyGridUpdate):
        """
        Update the grid given the occupancy grid update (applied on top of the current grid). Also update the flag for dstar to update the map.
        """

        self.map_lock.acquire()
        rospy.loginfo("[Dstar node]: Got map update")

        # print("\tgot an update")
        data_arr = np.array(data.data)

        if (np.all(data_arr == 0)):
            self.map_lock.release()
            return

        if (data_arr.shape[0] == 0) or (self.map.shape[0] == 0):
            self.map_lock.release()
            return

        temp_map = self.map.copy()

        index = 0
        for i in range(data.y, data.y + data.height):
            for j in range(data.x, data.x + data.width):
                temp_map[i][j] = data_arr[index]
                index += 1

        self.map = temp_map.copy()
        self.grid_update_needed = True

        self.map_lock.release()


    def position_callback(self, data: Odometry):
        position = data.pose.pose.position

        coords = [position.x, position.y]
        self.pose = coords


    def goal_callback(self, data: PoseStamped):
        self.goal = [data.pose.position.x, data.pose.position.y]

        self.goal_update_needed = True


    def dstar_loop(self):
        """
        Main loop for dstar. Update / manage dstar and its data, and publish the path whenever a new one becomes available.
        """

        rospy.init_node("dstar_ros_script")

        odom_topic = rospy.get_param("/odom_topic")
        goal_topic = rospy.get_param("/nav_goal_topic")

        map_topic = rospy.get_param("/nav/map_topic")
        map_update_topic = rospy.get_param("/nav/map_update_topic")

        frequency = 10  # hz

        rospy.Subscriber(map_topic, OccupancyGrid, self.grid_callback)
        rospy.Subscriber(map_update_topic, OccupancyGridUpdate, self.grid_update_callback)
        rospy.Subscriber(odom_topic, Odometry, self.position_callback)
        rospy.Subscriber(goal_topic, PoseStamped, self.goal_callback)

        rate = rospy.Rate(frequency)

        while not rospy.is_shutdown():

            # Startup condition: once all data is available, create a new Dstar object and find the path
            if (self.dstar is None) and len(self.map) > 0 and len(self.pose) > 0 and len(self.goal) > 0:

                self.dstar = Dstar(self.goal, self.pose, self.map.copy(), self.resolution, self.x_offset, self.y_offset, self.occupancy_threshold)
                self.publish_path(self.dstar.find_path())
                self.goal_update_needed = False

                continue

            if self.dstar is not None:

                if self.goal_update_needed:
                    # If we've gotten a new goal, it is more efficient to reset the dstar data (as it is all based on goal location)
                    # so we do so and find a new path

                    self.dstar = Dstar(self.goal, self.pose, self.map.copy(), self.resolution, self.x_offset, self.y_offset, self.occupancy_threshold)
                    self.publish_path(self.dstar.find_path())
                    self.goal_update_needed = False
                    continue
                

                if self.grid_update_needed:
                    # If the map has changed, let dstar make the necessary updates.
                    # Update the position and map, then find and publish a new path

                    self.dstar.update_position(self.pose)

                    rospy.logdebug("Dstar grid update")
                    
                    self.publish_path(self.dstar.update_map(self.map.copy(), self.x_offset, self.y_offset))

                    self.grid_update_needed = False
                    continue


            rate.sleep()

    def publish_path(self, path_data: list):
        """
        Publish the calculated path using the class's path publisher.
        Reduces the size/complexity of the path using the path sampling rate
        """

        if (len(path_data) == 0):
            return

        rospy.logdebug("Dstar publishing path")

        if path_data == "same":
            return

        path_data = np.array(path_data)

        path: Path = Path()
        path.poses = []
        path.header.stamp = rospy.Time.now()
        path.header.frame_id = "odom"

        for index, point in enumerate(path_data):
            if index % self.path_sampling_rate == 0 or index == len(path_data) - 1:
                # Sample every <path sampling rate> points (+ the last one)

                path_pose = PoseStamped()
                path_pose.header.stamp = rospy.Time.now()
                path_pose.header.frame_id = "odom"

                path_pose.pose.position.x = point[0]
                path_pose.pose.position.y = point[1]

                rotation = quaternion_from_euler(0, 0, 0)
                path_pose.pose.orientation.x = rotation[0]
                path_pose.pose.orientation.y = rotation[1]
                path_pose.pose.orientation.z = rotation[2]
                path_pose.pose.orientation.w = rotation[3]

                path.poses.append(path_pose)


        self.path_publisher.publish(path)


if __name__ == "__main__":
    dstar_node = DstarNode()
    dstar_node.dstar_loop()

File: lunabot_nav/src/dstar_node.cpp
#include "dstar.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "std_msgs/msg/bool.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "map_msgs/msg/occupancy_grid_update.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "nav_msgs/msg/path.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include <thread>

using std::placeholders::_1;


class DstarNode
{
public:
  DstarNode()
  {
    node = rclcpp::Node::make_shared("dstar_node");
  }

  void dstar_loop()
  {
    // rclcpp::param::get("/odom_topic", odom_topic);
    // rclcpp::param::get("/nav_goal_topic", goal_topic);
    // rclcpp::param::get("/nav/map_topic", map_topic);
    // rclcpp::param::get("/nav/map_update_topic", map_update_topic);
    // rclcpp::param::get("/nav/dstar_node/path_sampling_rate", path_sampling_rate); // Take every<n-th> point from the path
    // rclcpp::param::get("/nav/global_path_topic", path_topic);
    // rclcpp::param::get("/nav/occ_threshold", occupancy_threshold);

    // rclcpp::QoS *qos = new rclcpp::QoS(rclcpp::KeepAll());
    int qos = 10;
    // odom_topic = "/rtabmap/odom";
    odom_topic = "/rtabmap/position";
    goal_topic = "/goal";
    map_topic = "/maps/costmap_node/global_costmap/costmap";
    map_update_topic = "/maps/costmap_node/global_costmap/costmap_updates";
    path_sampling_rate = 5;
    path_topic = "/nav/global_path";
    occupancy_threshold = 50;

    int frequency = 3; // hz

    auto map_sub = node->create_subscription<nav_msgs::msg::OccupancyGrid>(map_topic, qos, std::bind(&DstarNode::grid_callback, this, _1));
    auto map_update_sub = node->create_subscription<map_msgs::msg::OccupancyGridUpdate>(map_update_topic, qos, std::bind(&DstarNode::grid_update_callback, this, _1));
    auto odom_sub = node->create_subscription<geometry_msgs::msg::PoseStamped>(odom_topic, qos, std::bind(&DstarNode::position_callback, this, _1));
    auto goal_sub = node->create_subscription<geometry_msgs::msg::PoseStamped>(goal_topic, qos, std::bind(&DstarNode::goal_callback, this, _1));
    path_pub = node->create_publisher<nav_msgs::msg::Path>(path_topic, 10);
    auto planning_enabled_subscriber = node->create_subscription<std_msgs::msg::Bool>("/nav/planning_enabled", qos, std::bind(&DstarNode::enable_callback, this, _1));
    auto costmap_enabled_subscriber = node->create_subscription<std_msgs::msg::Bool>("/costmap_enabled", qos, std::bind(&DstarNode::costmap_enabled_callback, this, _1));

    rclcpp::Rate rate(frequency);
    while (rclcpp::ok())
    {
      if (!dstar_init && map_init && goal_init && pose_init)
      {
        RCLCPP_DEBUG(node->get_logger(), "INIT DSTAR");
        dstar = Dstar(goal, pose, map, resolution, x_offset, y_offset, occupancy_threshold);
        dstar_init = true;
        RCLCPP_DEBUG(node->get_logger(), "Found Path");
        publish_path(dstar.find_path());
        goal_update_needed = false;
        grid_update_needed = false; // dif from python, don't update map if not strictly needed
        continue;
      }

      if (dstar_init)
      {
        if (goal_update_needed)
        {
          RCLCPP_DEBUG(node->get_logger(), "Goal Updating");
          // If we've gotten a new goal, reset the dstar data and find a new path
          dstar = Dstar(goal, pose, map, resolution, x_offset, y_offset, occupancy_threshold);
          RCLCPP_DEBUG(node->get_logger(), "Found Path");
          publish_path(dstar.find_path());
          goal_update_needed = false;
          grid_update_needed = false; // dif from python, don't update map if not strictly needed

          continue;
        }

        if (grid_update_needed)
        {
          // If the map has changed, update Dstar with the new position and map, then find and publish a new path
          dstar.update_position(pose);

          RCLCPP_DEBUG(node->get_logger(), "Dstar grid update");

          publish_path(dstar.update_map(map, x_offset, y_offset));
          grid_update_needed = false;

          continue;
        }
      }
      rclcpp::spin_some(node);
      rate.sleep();
    }
  }

private:
  Dstar dstar;
  std::vector<std::vector<int>> map;
  real_world_point goal;
  real_world_point pose;
  float resolution;   // meters per grid cell
  float x_offset = 0; // real world pose of the point 0, 0 in the grid
  float y_offset = 0;

  bool goal_init = false;
  bool pose_init = false;
  bool map_init = false;
  bool dstar_init = false;
  bool grid_update_needed = false;
  bool goal_update_needed = false;
  int empty_paths = 0; // how many empty paths in a row have been published

  std::mutex map_lock; // NOTE: this mutex was needed in python, might not be needed in C++ because roscpp callbacks are single threaded

  std::shared_ptr<rclcpp::Node> node;

  std::shared_ptr<rclcpp::Publisher<nav_msgs::msg::Path_<std::allocator<void>>, std::allocator<void>>> path_pub;

  std::string odom_topic;
  std::string goal_topic;
  std::string map_topic;
  std::string map_update_topic;
  std::string path_topic;
  int path_sampling_rate;
  float occupancy_threshold;

  bool planning_enabled = true;
  bool costmap_enabled = true;

  void enable_callback(const std_msgs::msg::Bool &msg)
  {
    this->planning_enabled = msg.data;
  }

  void costmap_enabled_callback(const std_msgs::msg::Bool &msg)
  {
    this->costmap_enabled = msg.data;
  }

  // Updates the map given a new occupancy grid.Update the flag such that dstar will update the map.
  void grid_callback(const nav_msgs::msg::OccupancyGrid &data)
  {
    // RCLCPP_INFO(node->get_logger(), "Got Grid");
    map_lock.lock();

    bool map_ok = false;
    for (int i = 0; i < data.info.width * data.info.height; i++)
    {
      if (data.data[i] != 0)
      {
        map_ok = true;
        break;
      }
    }

    uint32_t width = data.info.width;
    uint32_t height = data.info.height;

    map = std::vector<std::vector<int>>(height, std::vector<int>(width));

    for (int i = 0; i < height; ++i)
    {
      for (int j = 0; j < width; ++j)
      {
        map[i][j] = data.data[i * width + j];
      }
    }

    map_init = true;

    resolution = data.info.resolution;
    x_offset = data.info.origin.position.x;
    y_offset = data.info.origin.position.y;

    if (!map_ok)
    { // if the map is all zeroes, dont't update dstar with it
      map_lock.unlock();
      return;
    }

    grid_update_needed = true;

    map_lock.unlock();
  }

  // Update the grid given the occupancy grid update (applied on top of the current grid). Also update the flag for dstar to update the map.
  void grid_update_callback(const map_msgs::msg::OccupancyGridUpdate &data)
  {
    // RCLCPP_INFO(node->get_logger(), "Got Grid Update");
    map_lock.lock();

    bool map_ok = false;
    for (int i = 0; i < data.width * data.height; i++)
    {
      if (data.data[i] != 0)
      {
        map_ok = true;
        break;
      }
    }
    if (!map_ok)
    {
      map_lock.unlock();
      return;
    }

    int index = 0;
    for (int i = data.y; i < data.y + data.height; i++)
    {
      for (int j = data.x; j < data.x + data.width; j++)
      {
        if (0 <= i && i < map.size() && 0 <= j && j < map[0].size())
        { // prevents issue where dstar still processing and new map is loaded in
          map[i][j] = data.data[index];
          index++;
        }
      }
    }

    map_init = true;
    grid_update_needed = true;

    map_lock.unlock();
  }

  void position_callback(const geometry_msgs::msg::PoseStamped &data)
  {
    // RCLCPP_INFO(node->get_logger(), "Got Pose");
    pose.x = data.pose.position.x;
    pose.y = data.pose.position.y;
    pose_init = true;
  }

  void goal_callback(const geometry_msgs::msg::PoseStamped &data)
  {
    RCLCPP_DEBUG(node->get_logger(), "Got Goal");
    goal.x = data.pose.position.x;
    goal.y = data.pose.position.y;
    goal_update_needed = true;
    goal_init = true;
  }

  // Publish the calculated path using the class's path publisher.
  // Reduces the size / complexity of the path using the path sampling rate
  void publish_path(const std::vector<real_world_point> &path_data)
  {
    RCLCPP_DEBUG(node->get_logger(), "Publishing path");
    nav_msgs::msg::Path path;
    // rclcpp::Clock::SharedPtr clock = node.get_clock()
    path.header.stamp = rclcpp::Time();
    path.header.frame_id = "odom"; // Set the frame of reference

    if (path_data.size() == 0)
    {
      // Only publish some empty paths, avoid overwhelming traversal.
      if (empty_paths % 20 != 0)
      {
        return;
      }

      empty_paths++;
    }
    else
    {
      // when it's not empty, restart this counter.
      empty_paths = 0;
    }

    for (size_t index = 0; index < path_data.size(); ++index)
    {
      if (index % path_sampling_rate == 0 || index == path_data.size() - 1)
      {
        // Sample every <path_sampling_rate_> points (+ the last one)
        geometry_msgs::msg::PoseStamped path_pose;
        path_pose.header.stamp = rclcpp::Time();
        path_pose.header.frame_id = "odom";

        path_pose.pose.position.x = path_data[index].x;
        path_pose.pose.position.y = path_data[index].y;
        path_pose.pose.position.z = 0.0; // Assuming 2D path, Z is 0

        // Assuming no rotation (quaternion identity)
        path_pose.pose.orientation.x = 0;
        path_pose.pose.orientation.y = 0;
        path_pose.pose.orientation.z = 0;
        path_pose.pose.orientation.w = 1;

        path.poses.push_back(path_pose);
      }
    }

    bool mapAllZero = true;
    for (int i = 0; i < this->map.size(); i++)
    {
      for (int j = 0; j < this->map[i].size(); j++)
      {
        if (this->map[i][j] != 0)
        {
          mapAllZero = false;
          break;
        }
      }
    }

    // if the map is all zeroes, and the costmap is still on, don't publish paths
    if (mapAllZero && this->costmap_enabled)
    {
      return;
    }

    if (this->planning_enabled)
    {
      path_pub->publish(path);
    }
  }
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  DstarNode dstar_node;
  dstar_node.dstar_loop();
  return 0;
}
